contracts/Router.sol

```diff
@@ -0,0 +1,133 @@
+    event PayeeRemoved(address account, uint256 shares);
+    event PayeeShareAdjusted(address account, uint256 shares, uint256 oldShares);
+
+    mapping(address => uint256) private _shares;
+    mapping(address => mapping(address => uint256)) private _released;
+    address[] public _payees;
+
+    constructor(address[] memory payees, uint256[] memory shares_) payable {
+        require(payees.length == shares_.length, "Router: payees and shares length mismatch");
+        require(payees.length > 0, "Router: no payees");
+
+        for (uint256 i = 0; i < payees.length; i++) {
+            _addPayee(payees[i], shares_[i]);
+        }
+    }
+
+    // ---------------------------- getters ----------------------------
+    function shares(address account) public view returns (uint256) {
+        return _shares[account];
+    }
+
+
+    function released(address account, address token) public view returns (uint256) {
+        return _released[token][account];
+    }
+
+    function totalShares() public view returns (uint256) {
+        return _totalShares;
+    }
+
+    function totalReleased(address token) public view returns (uint256) {
+        return _totalReleased[token];
+    }
+    // ---------------------------- external functions ----------------------------
+    function release(IERC20 token) external nonReentrant {
+        require(_totalShares > 0, "Router: no shares");
+        uint256 balance = token.balanceOf(address(this));
+        uint256 total = 0;
+        for(uint256 i = 0; i < _payees.length; i++) {
+            address payee = _payees[i];
+            uint256 payment;
+            if (i == _payees.length - 1){
+                payment = balance - total;
+            } else {
+                payment = balance * _shares[payee] / _totalShares;
+            }
+            if (payment > 0) {
+                _released[payee][address(token)] = _released[payee][address(token)] + payment;
+                SafeERC20.safeTransfer(token, payee, payment);
+                total += payment;
+            }
+        }
+        _totalReleased[address(token)] = _totalReleased[address(token)] + total;
+        emit PaymentReleased(token, total);
+    }
+
+     function addPayee(address account, uint256 shares_) external onlyOwner {
+        _addPayee(account, shares_);
+    }
+
+    function removePayee(address account) external onlyOwner {
+        _removePayee(account);
+    }
+
+    function adjustShare(address account, uint256 shares_) external onlyOwner {
+        _adjustShare(account, shares_);
+    }
+
+
+    
+    // ---------------------------- private functions ----------------------------
+
+    function _addPayee(address account, uint256 shares_) private {
+        require(account != address(0), "Router: account is the zero address");
+        require(shares_ > 0, "Router: shares are 0");
+        require(_shares[account] == 0, "Router: account already has shares");
```

Comment:
We're not currently testing that this fails if the account already has shares. We should add a test for that